text/0000-xor-urls/0000-xor-urls.md

```diff
@@ -0,0 +1,115 @@
+
+### XOR-URLs specification
+The following are the main requirements for the encoding to be used to generate the XOR-URLs:
+
+- Be able to support new and different types of base encodings and hash functions for the XOR addresses in the future.
+- Include the content type within the XOR URL which would allow the client app to correctly render the data to the user especially when referencing an ImmutableData.
+
+It is proposed the use of [multiformats](https://github.com/multiformats) and [CID](https://github.com/ipld/cid), which allow us to cover the above requirements. A CID identifier can be used in the XOR-URL for specifying the XOR address part, and have additional parts to support MutableData’s type tag, version, as well as the path and query parts. SAFE URLs can then be defined in the following way (BNF 1-like):
+```
+<safe-url>            := 'safe://' ( <xor-uri> | <public-name-uri> )
+
+<public-name-uri>     :=  [<service> '.'] <public-name> <path-query-fragment>
+
+<xor-uri>             := <immutable-data-uri> | <mutable-data-uri>
+
+<immutable-data-uri>  := <cid> <query-fragment>
+
+<mutable-data-uri>    := <cid> ':' <type-tag> ['+' <content-version>] <path-query-fragment>
+
+<path-query-fragment> := ['/' <path>] <query-fragment>
+
+<query-fragment>      := ['?' <query>] ['#' <fragment>]
+```
+
+Where:
+
+- `<cid>`: follows the [CID format](https://github.com/ipld/cid) which self-describes and encodes:
+    - the base encoding for the string, it's here proposed to use [z-base32 encoding](http://philzimmermann.com/docs/human-oriented-base-32-encoding.txt) for the reasons explained above,
+    - the version of CID for upgradability purposes (v1 now),
+    - the content type or codec of the data being referenced,
+    - the XOR address of the content being referenced
+- `<content-version>`: for future implementation to reference versionable content, using a single address with different versions
+
+- `<type-tag>`: the type tag value if the CID is referencing a MutableData. In the absence of this value, the CID will be assumed to be targetting an ImmutableData
+
+- `<path>`: the path of the file if the CID is referencing a MutableData, which can be accessed through the NFS/Files convention (or other emulations/conventions in the future) and resolve it
+
+- `<query>`: query arguments, to be used by the client app and not for retrieving the content
+
+- `<fragment>`: fragment of the content, to be used by the client app and not for retrieving the content
+
+The resolver function shall first attempt to decode the URL assuming it contains a `<cid>` part. If that step fails (either because it couldn’t decode the CID, or because there was no data at the decoded XOR address), it will do a fallback to assume it’s a public-name URL and it shall try to find the pulic-name MD (with the `sha3` hash of the string), just like any public-name URLs are resolved now by the resolver function.
```

Comment:
`The resolver function shall first attempt to decode the URL assuming it contains a `<cid>` part.`

This is slightly different to the PNS RFC atm. If this is the desired process, we should be sure to sync up the two. (more a note to self :P )

Reply:
@joshuef I don't see a difference really, can you elaborate?
Perhaps the only difference is that I'm saying we should try to decode the CID and attempt to fetch it before falling back, rather than only decoding, is this what you see different?
The reason I'm proposing that is to not eclipse a public name which happens to be a valid CID, it will be eclipsed anyway if there exist content at the location encoded by the CID but at least not do it when there is no content. But this again is what's proposed here, so it's open for debate.

Reply:
I'm adding this explanation to that paragraph as it may help

Reply:
Ah you're right. The part I'm talking about is discerning uris vs other graph entries when referencing from `default`. Sorry for confusion!